citest/gcp_testing/google_idp_session.py

```diff
@@ -0,0 +1,195 @@
+    the request/response headers.
+  * Possibly, a Http session is established with the third party application
+    as well.
+  * Make requests in the established session scope.
+  * Log out after the necessary exchange is made.
+
+  This class automtes both the Google account login and scope approval to create
+  an authenticated HTTP session, and provides access to the session cookies for
+  use with urllib2 Requests. Both direct access to the Cookies and mutating
+  requests via RequestMutator methods are supported.
+
+  Example usage:
+    # target_url should forward to 'accounts.google.com/ServiceLogin' page
+    # via an OAuth2 or SAML redirect.
+    target_url = "http://<gate_host>/credentials"
+    logout_url = "http://<gate_host>/auth/logout"
+    session = GoogleIDPSession(target_url,
+                           "user@domain.net",
+                           "secret_pass_word",
+                           logout_url=logout_url)
+    session.initialize()
+    req = urllib2.Request(url=target_url)
+    self.mutate_request(req) # loads cookies into request header
+    resp = urllib2.urlopen(req)
+    session.terminate()
+  """
+
+
+  def __init__(
+      self, target_url, login, password,
+      user_approval_url='https://accounts.google.com/ServiceLoginAuth',
+      logout_url=None):
+    """Construct an instance.
+
+    Args:
+      target_url [String]: A URL that will eventually redirect to
+        'https://accounts.google.com/ServiceLogin' given a new session.
+      login [String]: Username/email to login as.
+      password: [String] Password for the provided login.
+      user_approval_url [String]: (Optional) The URL for a Google user to grant
+        approval to an application. Can override if necessary.
+      logout_url [String]: (Optional) The logout url of the application we are
+        authenticating to.
+    """
+    self.__session = requests.session()
+    self.__target_url = target_url
+    self.__login = login
+    self.__password = password
+    self.__user_approval_url = user_approval_url
+    self.__logout_url = logout_url
+    self.__is_initialized = False
+
+
+  def initialize(self):
+    """Initialize a session using Google as the IDP."""
+    if self.__is_initialized:
+      return
+
+    try:
+      login_resp = self.__session.get(self.__target_url)
+    except Exception as e:
+      __throw_http_session_exception(e, 'GET', self.__target_url)
+
+    if (GOOGLE_ACCOUNT_LOGIN_URL
+        not in login_resp.url[:len(GOOGLE_ACCOUNT_LOGIN_URL)]):
+      raise idp_manager.IDPSessionError(
+        'Target URL did not redirect to Google account login page.'
+      )
+    login_form_inputs = (BeautifulSoup(login_resp.content)
+        .find('form')
+        .find_all('input'))
+    login_dict = {}
+    for input in login_form_inputs:
+      if input.has_attr('value'):
+        login_dict[input['name']] = input['value']
+
+    login_dict['Email'] = self.__login
```

Comment:
If these are in the input you are overriding them. Is that intentional?
Otherwise initlalize login_dict above as {'Email': self.__login, 'Passwd': self.__password}